src/util/mpir_hwtopo.c

```diff
@@ -554,3 +554,61 @@ uint64_t MPIR_hwtopo_get_node_mem(void)
 
     return size;
 }
+
+bool MPIR_hwtopo_is_dev_close_by_name(const char *name)
+{
+    bool is_close = false;
+    if (!bindset_is_valid)
+        return is_close;
+#ifdef HAVE_HWLOC
+    MPIR_hwtopo_gid_t gid = MPIR_hwtopo_get_obj_by_name(name);
+    int hwloc_obj_index = HWTOPO_GET_INDEX(gid);
+    int hwloc_obj_depth = HWTOPO_GET_DEPTH(gid);
+    hwloc_obj_t first_non_io =
+        hwloc_get_obj_by_depth(hwloc_topology, hwloc_obj_depth, hwloc_obj_index);
+    /* Determine if PCI device is "close" to process by checking if process's affinity is included
+     * in PCI device's affinity or if PCI device's affinity is included in rank's affinity */
+    is_close = (hwloc_bitmap_isincluded(bindset, first_non_io->cpuset) ||
+                hwloc_bitmap_isincluded(first_non_io->cpuset, bindset));
```

Comment:
How about neither but overlap?

Reply:
I'm not sure what you mean here.

Reply:
A rather artificial example --
   nic1 -- 0, 1
   nic2 -- 4, 5

Should we say process bind to 1,2 is close to nic1?

Reply:
Yes. It will find the NIC attached to your socket.

Reply:
I think you didn't quite get my example. In my example, the `bindset` is `1, 2`, the `non_io->cpuset` is either `0, 1` or `4, 5`. so neither `hwloc_bitmap_isincluded` will return true, thus you won't find the cloeses nic.

Reply:
I _think_ I see what you're talking about now (this isn't originally my patch and i've not gone into hwloc details before).

I think in that case, no NIC would be listed as "close" because the process might not be on a core that is "close" to the NIC.

So then, that process might have no close NICs, which would trigger the code that falls back to using the other NICs.

Reply:
That's fine. Anyway, I think my main point is that this logic is not trivia and we need refactor this logic into a function so the maintainer can read this code separately with a cool head. And if later we want to tune the logic, we don't have to repeat the fix in multiple places.

Reply:
I think the function names are pretty clear now that I've moved things around (the code we're pointing to in this comment is already in a new function now). Let me know if it still looks odd to you.

If you follow the code path, you'll get a call stack like this:

```
setup_multi_nic()
is_nic_close()
MPIR_hwtopo_is_dev_close_by_pci()
pci_device_is_close()
```

Inside `setup_multi_nic`, there's quite a bit of documentation explaining the code path and what happens if a NIC is close or not.

Reply:
I think that's good enough. 👍